File: back-end/gateways/airo-users-gateway-api/src/profile/profile.service.ts
import { HttpService } from '@nestjs/axios';
import { Injectable } from '@nestjs/common';
import { firstValueFrom } from 'rxjs';

export type ProfileGetProfileByUidResponse = {
    uid: string,
    firstName: string,
    lastName: string,
    nickname: string,
    accountType: 'free' | 'pro',
};

@Injectable()
export class ProfileService {
    private readonly profileServiceUrl = process.env.PROFILE_API_URL!;

    constructor(private readonly httpService: HttpService) { }

    async createProfile(uid: string, accountType: string, email: string, nickname: string, creditCardDetails?: string): Promise<void> {
        await firstValueFrom(
            this.httpService.post(`${this.profileServiceUrl}/api/profile`, {
                uid,
                accountType,
                email,
                nickname,
                creditCardDetails,
            }),
        );
    }

    async getProfileByUid(uid: string): Promise<ProfileGetProfileByUidResponse> {
        const response = await firstValueFrom(
            this.httpService.get(`${this.profileServiceUrl}/api/profile?uid=${uid}`),
        );
        return response.data;
    }

    async updateProfileByUid(uid: string, firstName: string, lastName: string): Promise<ProfileGetProfileByUidResponse> {
        const response = await firstValueFrom(
            this.httpService.patch(`${this.profileServiceUrl}/api/profile`, { uid, firstName, lastName }),
        );
        return response.data;
    }
}

File: back-end/microservices/airo-profile-microservice/Services/Impl/ProfileService.cs
using airo_profile_microservice.Models;
using airo_profile_microservice.Services.Interfaces;

namespace airo_profile_microservice.Services.Impl;

public class ProfileService(IProfileRepository profileRepository) : IProfileService
{
    public async Task CreateProfile(string uid, string accountType, string email, string? creditCardDetails)
    {
        await profileRepository.SetProfile(uid, new Dictionary<string, object>
        {
            { "uid", uid },
            { "firstName", "" },
            { "lastName", "" },
            { "accountType", accountType },
            { "email", email },
            { "creditCardDetails", creditCardDetails ?? ""},
        });
    }

    public async Task UpdateProfile(string uid, string firstName, string lastName)
    {
        await profileRepository.SetProfile(uid, new Dictionary<string, object>
        {
            { "firstName", firstName },
            { "lastName", lastName }
        });
    }

    public Task<Profile?> GetByUid(string uid) => profileRepository.GetProfile(uid);
}
